### ISSUE TITLE
Declarative role management (`Role` CRD)

### ISSUE THREAD

**Issue by gbartolini** [cloudnative-pg/cloudnative-pg#5341]
With the implementation of database management via the `Database` CRD, it’s important to offer an alternative approach to role management, currently handled through the `.spec.managed.roles` stanza. We propose introducing a new CRD called `Role`, which will eventually replace the existing implementation. For now, we will maintain both interfaces, ensuring that they are mutually exclusive if coexistence is not feasible.

**Comment by mnencia** [comment]
Given the current timeline and the need to release version 1.25 before the end of the year, there is not enough time to complete this feature within the 1.25 release cycle. I’m moving this to version 1.26 for better planning and execution.

**Comment by mkris98** [comment]
Having a possibility for specifying pgaudit.log on the user to enable session level auditing would be very useful too

**Comment by gbartolini** [comment]
> Moving to release 1.27

@mnencia, can you please provide the reasons for this decision for transparency? Thanks!

**Comment by gbartolini** [comment]
Considering our application for the Linux Foundation Mentorship program (#4683), our main objective to introduce support for the dynamic loading of Postgres extensions (issue #7188) and the isolation checker (issue #7466), I recommend that we postpone role management to version 1.28 for now. I would prefer a shorter cycle for version 1.27 that includes the two aforementioned features, which are ready for Postgres 18, rather than extending the timeline like we did for version 1.26 due to the numerous important features and changes included in that release.

I know it is a difficult decision, given how I think this feature is critical for the future of CNPG, but I'd rather under-commit for 1.27. What do the other maintainers think?

**Comment by leonardoce** [comment]
I'm for postponing to 1.28.  We already have many interesting things for 1.27.

**Comment by mnencia** [comment]
> Considering our application for the Linux Foundation Mentorship program ([#4683](https://github.com/cloudnative-pg/cloudnative-pg/issues/4683)), our main objective to introduce support for the dynamic loading of Postgres extensions (issue [#7188](https://github.com/cloudnative-pg/cloudnative-pg/issues/7188)) and the isolation checker (issue [#7466](https://github.com/cloudnative-pg/cloudnative-pg/pull/7466)), I recommend that we postpone role management to version 1.28 for now. I would prefer a shorter cycle for version 1.27 that includes the two aforementioned features, which are ready for Postgres 18, rather than extending the timeline like we did for version 1.26 due to the numerous important features and changes included in that release.
> 
> I know it is a difficult decision, given how I think this feature is critical for the future of CNPG, but I'd rather under-commit for 1.27. What do the other maintainers think?

I agree. Let's plan for a shorter release cycle and move this feature to 1.28.

**Comment by fcanovai** [comment]
I agree too. 1.28.

**Comment by ShadowJonathan** [comment]
????????????

**Comment by ShadowJonathan** [comment]
@gbartolini, @mnencia, @leonardoce the AI just closed this as not-planned, could you please re-open this?

**Comment by guesant** [comment]
why it was moved to not planned?

**Comment by jsilvela** [comment]
OK, though I'm not a maintainer I seem to have permissions to reopen this.
I see the `no-stale` tag was applied, but that didn't remove the `stale` flag that had been applied by github-actions.
Would be good for @mnencia @gbartolini to clarify if this is in the plans or not

**Comment by gbartolini** [comment]
I dislike bringing this up, but given our current workload for version 1.28, I genuinely cannot see how we can tackle this feature and complete it in the next 6 to 7 weeks. I propose to postpone this feature to the next release once again. What are your thoughts?

**Comment by gbartolini** [comment]
> Would be good for [@mnencia](https://github.com/mnencia) [@gbartolini](https://github.com/gbartolini) to clarify if this is in the plans or not

This is an important feature that is definitely planned. We just need to find a way to adhere to the scheduled release in the roadmap. Having an existing method for managing roles does not increase the demand for this feature. It is also too complex for a mentorship project - which is what we are currently focusing on at the moment.

**Comment by jsilvela** [comment]
You are aware though, that there is already quite a bit of this feature coded?
And some problems associated to the new CRD approach were identified, that have been discussed and had proposed solutions.
Would be great if the maintainers could do a reconnaissance, perhaps even review the code, and then get a better idea of the amount of work remaining (which is not insignificant).

**Comment by ShadowJonathan** [comment]
> I propose to postpone this feature to the next release once again. What are your thoughts?

This is fine (imo), my only concern was that an AI bot mistakenly and suddenly closed this issue without warning or notice, when this is actively planned, as you said.

**Comment by gbartolini** [comment]
> You are aware though, that there is already quite a bit of this feature coded?

Of course, I am aware, but I have overlooked the PR due to other priorities and features.

> And some problems associated to the new CRD approach were identified, that have been discussed and had proposed solutions. Would be great if the maintainers could do a reconnaissance, perhaps even review the code, and then get a better idea of the amount of work remaining (which is not insignificant).

I am concerned there's not enough time between now and the planned date for 1.29 to fully focus on this matter, without causing context switch pains (which we should avoid, especially now). I would like to hear the other maintainers' PoV though.

**Comment by jsilvela** [comment]
To be clear: I'm not advocating for this work to be shipped in 1.28, or in 1.29.
I just want the work already done not to be forgotten.
And as a summary of the current state in case it was beginning to be forgotten:

- the code already has a working controller plus role CRD (named PgRole to avoid collisions with the K8s Role)
- two main things remain to be implemented:
  - the code to grant k8s permissions to the PgRole objects is a bit hacky, more involved than for existing ManagedRoles. There was a proposal from Leonardo to simplify by creating RBAC objects per PgRole and relying on garbage collection to reclaim
  - how to manage co-existence or deprecation of managed roles with CRD-based roles. There was a proposal by Armando I believe to use webhooks to transform managed roles into CRD-based roles on the fly and have only one controller/flow

**Comment by gbartolini** [comment]
Understood. Of course, it seems reasonable for me that, once this becomes a priority, maintainers will start reviewing what's already been done and, at that point, decide if and what to keep from that work. For now, I want the maintainers' input and approval to move this to 1.29. Thanks.

**Comment by mnencia** [comment]
I agree to postpone it.

**Comment by leonardoce** [comment]
Of course we should not forget this work. Unfortunately we don't have enough time to address this for 1.28, but we should to it for 1.29. Let's postpone it.

**Comment by sxd** [comment]
Agree, we should move it to 1.29, and start working on reviewing the code as soon as we can, because this is going to take sometime and thinking about having a beta in 1.29 of this, and hopefully have it stable in 1.30.

**Comment by gbartolini** [comment]
With majority of the maintainers, moved to 1.29. Thanks.

**Comment by gbartolini** [comment]
My suggestion is to use, for consistency, the `Role` custom resource, rather than `PGRole` (see my comment](https://github.com/cloudnative-pg/cloudnative-pg/pull/6155/files#r2640836643)).

As far as the co-existence of the `Role` CR with the current `.spec.managed.roles` capability is concerned, there can be three use cases regarding a specific role:

1. The role is created via `.spec.managed.roles`
2. The role is created via the new `Role` CR
3. The role is created directly in PostgreSQL (with a user having superuser/role privileges)

Let's skip the third use case for now and focus on the first two.

A role can only be reconciled by either 1 or 2, with 1 having higher precedence given that it is the current behaviour.

For example, if you create the `dante` user as follows:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
spec:
  managed:
    roles:
    - name: dante
      ensure: present
      comment: Dante Alighieri
      login: true
      superuser: false
      inRoles:
        - pg_monitor
        - pg_signal_backend
```

and then you create the following role object:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Role
metadata:
  name: cluster-example-dante
spec:
  cluster:
    name: cluster-example
  name: dante
  ensure: present
  comment: Dante Alighieri
  login: true
  superuser: false
  inRoles:
    - pg_monitor
    - pg_signal_backend
  <snip>
```

The latter will not be reconciled by CNPG, resulting in an error or a skipped result.

If you swap the order, the result will not change: the `managed` one will have precedence over the `role` resource, leading to the same error.

But we should enable, in this first implementation, the coexistence of the two methods for distinct roles.

For example, you could create the `dante` role via the `managed` stanza, and the `boccaccio` one via the `role` CR.

Obviously, we should then converge in terms of recommendations (`role` CR) and also in terms of implementation (by reusing the same code).

Thoughts?

**Comment by gbartolini** [comment]
I was discussing a scenario with @leonardoce, and I really liked his idea of adding a command to the plugin that simplifies the migration from the `managed` case to the `role` case. For example:

```sh
kubectl cnpg convert managed-to-crds  [--dry-run]
```

This command would automatically take all the roles created by CNPG (such as `app`) and those defined in the `managed` sections to generate the necessary `Role` resources. It would also create the application `Database` resource and suggest a new `Cluster` resource without the `managed.roles` section.

Implementing this feature would allow us to disable, in the future, the default creation of the `app` user and the `app` database.

P.S.: Feel free to propose a better name and interface.

**Comment by jsilvela** [comment]
On `PgRole` vs `Role`. That was my proposal, and nobody in the team *liked* it, including me.
But using `Role` is a footgun. When I was doing testing of the implementation (what we had so far) I kept stumbling on the footgun, and I thought, wait until CNGP newbies, or Kubernetes newbies, get to use this, and just anticipate the cries for help in Slack and support calls.
But yeah, I mean, I agree, if we use `PGRole` shouldn't we use `PG` as a prefix for other stuff? It's a good point. But those other things don't have such a direct conflict.

I like the plugin proposal. 👍

**Comment by NHellFire** [comment]
> My suggestion is to use, for consistency, the `Role` custom resource, rather than `PGRole` ([see my comment](https://github.com/cloudnative-pg/cloudnative-pg/pull/6155/files#r2640836643)).

As a user, I'd prefer `PGRole` to avoid needing to specify the full name. Ideally the other CRs should also be prefixed with `PG` for consistency but those at least (currently) don't clash with Kubernetes's own names.

**Comment by svenakela** [comment]
> On `PgRole` vs `Role`. That was my proposal, and nobody in the team _liked_ it, including me. But using `Role` is a footgun. When I was doing testing of the implementation (what we had so far) I kept stumbling on the footgun, and I thought, wait until CNGP newbies, or Kubernetes newbies, get to use this, and just anticipate the cries for help in Slack and support calls. But yeah, I mean, I agree, if we use `PGRole` shouldn't we use `PG` as a prefix for other stuff? It's a good point. But those other things don't have such a direct conflict.
> 
> I like the plugin proposal. 👍

What about `DatabaseRole`?
https://www.postgresql.org/docs/current/database-roles.html

**Comment by los93sol** [comment]
This would be a great addition to allow apps to declare their roles along with the database when a cluster is shared.  It would also be perfect if it would create the secret in the app namespace.

**Comment by gbartolini** [comment]
> What about `DatabaseRole`? https://www.postgresql.org/docs/current/database-roles.html

Wow! I hadn't thought about that and had completely overlooked the Postgres docs page. I like it.

**Comment by gbartolini** [comment]
I have changed the title to use the `DatabaseRole` name for the CRD.

**Comment by gbartolini** [comment]
Update. The proposal is to name the CRD `DatabaseRole` ([see @svenakela's comment](https://github.com/cloudnative-pg/cloudnative-pg/issues/5341#issuecomment-3691746163)).

As far as the co-existence of the `DatabaseRole` CR with the current `.spec.managed.roles` capability is concerned, there can be three use cases regarding a specific role:

1. The role is created via `.spec.managed.roles`
2. The role is created via the new `Role` CR
3. The role is created directly in PostgreSQL (with a user having superuser/role privileges) - outside the scope.

Let's focus on the first two.

A role can only be reconciled by either 1 or 2, with 1 having higher precedence (since it is the current behaviour).

For example, if you create the `dante` user as follows:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
spec:
  managed:
    roles:
    - name: dante
      ensure: present
      comment: Dante Alighieri
      login: true
      superuser: false
      inRoles:
        - pg_monitor
        - pg_signal_backend
```

and then you create the following role object:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: DatabaseRole
metadata:
  name: cluster-example-dante
spec:
  cluster:
    name: cluster-example
  name: dante
  ensure: present
  comment: Dante Alighieri
  login: true
  superuser: false
  inRoles:
    - pg_monitor
    - pg_signal_backend
  <snip>
```

The latter will not be reconciled by CNPG, resulting in an error or a skipped result.

If you swap the order, the result will not change: the `managed` one will have precedence over the `role` resource, leading to the same error.

But we should enable, in this first implementation, the coexistence of the two methods for distinct roles.

For example, you could create the `dante` role via the `managed` stanza, and the `boccaccio` one via the `role` CR.

The general direction is that we should then converge on recommendations (`role` CR) and on implementation (by reusing the same code).

**Comment by tarteo** [comment]
Hi, I would like this feature. Is there anyway I can help? Let me know

**Comment by gbartolini** [comment]
We are currently focused on completing our extensions story, which includes image catalog support for extensions (#8495), as well as the Postgres extension container images project (https://github.com/cloudnative-pg/postgres-extensions-containers/issues/15). Therefore, we need to make a decision regarding our timeline:

1. Should we postpone the release of version 1.29 to include this feature?
2. Or should we move this feature to version 1.30 and proceed with the release of version 1.29?

I vote for the second option: moving this feature to version 1.30.

**Comment by leonardoce** [comment]
I vote for the second option too.

**Comment by armru** [comment]
I vote for the second option. I prefer releasing the 1.29 sooner to bring fixes to the end users

**Comment by mnencia** [comment]
I agree to moving the feature to 1.30

**Comment by gbartolini** [comment]
Ok, I see the majority has voted. Moving this to 1.30.
